### PR TITLE
[spmd compile API] add a (temporary) mechanism for overriding input tensors' placements

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98180
* #98179
* #98178
* #98177

Currently, the compile API assumes all input tensors' shard dimension is the first dimension. dtensor expansion doesn't work when there are input tensors whose shard dimension is not the first dimension.

In addtion, respect non-tensor inputs beyond nn.Module and optim.Optimizers.

Differential Revision: [D44615234](https://our.internmc.facebook.com/intern/diff/D44615234/)

### CODE REVIEW

File: torch/distributed/_spmd/api.py [hunk 1]
```diff
@@ -153,10 +162,21 @@ def _dtensor_expand(
     for a in flat_args:
         if isinstance(a, torch.Tensor):
             inps.append(a)
-            schemas.append(shard_schema)
-        elif isinstance(a, nn.Module) or isinstance(a, torch.optim.Optimizer):
-            # nn.Module or optimizer placeholder is captured by make_fx but
-            # never used in the graph
+            if id(a) in _placements_override:
```
Comment (mrshenli) [inline torch/distributed/_spmd/api.py:165]: TODO for @mrshenli: investigate if there is a better way to handle these overrides.

**Review by mrshenli — APPROVED** [review]
stamp to unblock

**Review by kumpera — APPROVED** [review]
LGTM

**Comment by mrshenli** [comment]
superseded by #98232

**Pull request closed without merging**